sdk/python/lib/test/test_resource.py

```diff
@@ -296,6 +306,46 @@ def __init__(self, name: str, opts=None):
         depends_on=[a.b]
     ))
 
+# Regression test for https://github.com/pulumi/pulumi/issues/13551
+@pulumi.runtime.test
+def test_complex_implicit_parent_child_dependencies():
+    mocks.set_mocks(MinimalMocks())
+    
+    class A(pulumi.ComponentResource):
+        def __init__(self, name: str, opts=None):
+            super().__init__("my:modules:A", name, {}, opts)
+            a_1 = pulumi.CustomResource("my:module:Child-1", "a-child-1", opts=ResourceOptions(parent=self, depends_on=[self]))
+            pulumi.CustomResource("my:module:Child-2", "a-child-2", props={"transitive_urn": a_1.urn} ,opts=ResourceOptions(parent=self))
+
+    class B(pulumi.ComponentResource):
+        def __init__(self, name: str, opts=None):
+            super().__init__("my:modules:B", name, {}, opts)
+            b_1 = pulumi.CustomResource("my:module:Child-1", "b-child-1", opts=ResourceOptions(depends_on=[self]))
+            pulumi.CustomResource(
+                "my:module:Child-2",
+                "b-child-2",
+                props={"transitive_urn": b_1.urn},
+                opts=ResourceOptions(depends_on=[self])
+            )
+
+    class C(pulumi.ComponentResource):
+        def __init__(self, name: str, opts=None):
+            super().__init__("my:modules:C", name, {}, opts)
+            c_1 = pulumi.CustomResource("my:module:Child-1", "c-child-1", opts=ResourceOptions(parent=self))
+            pulumi.CustomResource(
+                "my:module:Child-2",
+                "c-child-2",
+                props={"transitive_urn": c_1.urn},
+                opts=ResourceOptions(parent=self)
+            )
+
+
+    # with pytest.raises(RuntimeError):
+    #     A("a") # Fails
```

Comment:
I would love some guidance on how to re-enable this assertion, as it's essentially the core of this change. Unfortunately, the exception is raised within a process that appears to not propagate back up to this process (within an asyncio task in `rpc`) which means the "standard" method of catching exceptions fails to do so here.

Reply:
I've seen mentions of some `asyncio` hacking other places - is there a way to get a handle on the event loop that will eventually run the dependency functions and assert an error is thrown there?

Reply:
Fallback is to just write this as an integration test in test/integration and check the program exits with the expected error